src/layoutBuilder.js

```diff
@@ -388,75 +495,288 @@ function convertWordsToCodepoints(lineAsArrayOfCodepoints) {
+		} else if (index + 1 === lineAsArrayOfCodepoints.length) {
+			const direction = getWordDirection(currentWord);
+			currentWord.push(lineAsArrayOfCodepoints[index]);
+			// similar to above, if we hit the end of a line (or a newline character)
+			// we need to add a whitespace codepoint
+			if (direction === 'RTL') {
+				currentWord.unshift(whiteSpaceAsCodePoint);
+			}
+			arrayOfCodePoints.push(currentWord);
+			currentWord = [];
+		} else {
+			currentWord.push(lineAsArrayOfCodepoints[index]);
+		}
+	}
+	return arrayOfCodePoints;
+}
+
+function getWordDirection(codePoints) {
+	const arabic = [0x0600, 0x06ff];
+	const arabicSupplement = [0x0750, 0x077f];
+	const arabicExtendedA = [0x08a0, 0x08ff];
+	const arabicPresentationFormsA = [0xfb50, 0xfdff];
+	const arabicPresentationFormsB = [0xfe70, 0xfeff];
+	const hebrew = [0x0590, 0x05ff];
+
+	const testCases = [
+		arabic,
+		arabicSupplement,
+		arabicExtendedA,
+		arabicPresentationFormsA,
+		arabicPresentationFormsB,
+		hebrew,
+	];
+
+	if (
+		codePoints.some((codePoint) => {
+			return testCases.some(([testLow, testHigh]) => {
+				return testLow <= codePoint && codePoint <= testHigh;
+			});
+		})
+	) {
+		return 'RTL';
+	} else {
+		return 'LTR';
+	}
+}
+
+/**
+ * Takes the inlines from the current line. Reorders the inlines using BIDI. After reordering
+ * the inlines it uses them to make a new line and returns that line
+ * @param {Line} line The original line that contains some inline RTL
+ * @param {StyleContextStack} styleStack The style context used when rebuilding the inlines
+ * @param {TextTools} textTools Used to rebuild the inlines
+ * @param {Object} textNode The original textNode that contains the nested inline rtl text
+ * @param {Number} availableWidth the line width being used
+ * @returns {Line}
+ */
+function addLineWithInlineRTL(
+	line,
+	styleStack,
+	textTools,
+	textNode,
+	availableWidth
+) {
+	// line.inlines contains every word that will appear on this line in an an array of objects
+	// We want to use these to generate a lookup so that we can reassign the fonts/styles after
+	// we've reordered the words in the line
+	// NB there is an inherent problem with this - repeated words in the line will
+	// only get one entry in the lookup, so styling may not pass through completely correctly
+	const wordPropsLookup = {};
+
+	// CJK gets split up into individual characters in the pdfmake generated line, so we need
+	// to deal with this differently as the bidiWord we get back later comes back as a string
+	// of CJK characters
+	let CJKbuffer = { text: '', font: 'NotoSansCJK', style: '' };
+	line.inlines.forEach((word, index, array) => {
+		// extract the parts we're interested in, and to avoid whitespace woes we will
+		// key using the trimmed text
+		const isLastWord = index === array.length - 1;
+		const lastCharWasSpace = word.text.slice(-1) === ' ';
+		const text = word.text.trim();
+		const font = word.font.name.split('-')[0].replace('TCRegular', '');
```

Comment:
What does TCRegular refer to here?

Reply:
Annoyingly the Chinese font was misnamed - all this nastiness was coming about from running bidi on the whole line instead of just the RTL. 

I've refactored and removed this anyhow.